programs/mango-v4/src/instructions/perp_liq_base_and_positive_pnl.rs

```diff
@@ -0,0 +1,434 @@
+        return Ok(());
+    }
+
+    let mut perp_market = ctx.accounts.perp_market.load_mut()?;
+    let perp_market_index = perp_market.perp_market_index;
+    let settle_token_index = perp_market.settle_token_index;
+    let base_lot_size = I80F48::from(perp_market.base_lot_size);
+
+    let mut settle_bank = ctx.accounts.settle_bank.load_mut()?;
+    // account constraint #2
+    require!(
+        settle_bank.token_index == settle_token_index,
+        MangoError::InvalidBank
+    );
+
+    // Get oracle price for market. Price is validated inside
+    let oracle_price = perp_market.oracle_price(
+        &AccountInfoRef::borrow(ctx.accounts.oracle.as_ref())?,
+        None, // checked in health
+    )?;
+    let price_per_lot = cm!(base_lot_size * oracle_price);
+
+    // Fetch perp positions for accounts, creating for the liqor if needed
+    let liqee_perp_position = liqee.perp_position_mut(perp_market_index)?;
+    require!(
+        !liqee_perp_position.has_open_taker_fills(),
+        MangoError::HasOpenPerpTakerFills
+    );
+
+    let liqor_perp_position = liqor
+        .ensure_perp_position(perp_market_index, perp_market.settle_token_index)?
+        .0;
+
+    // Settle funding
+    liqee_perp_position.settle_funding(&perp_market);
+    liqor_perp_position.settle_funding(&perp_market);
+
+    // Max settleable on the liqee?
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    liqee_perp_position.update_settle_limit(&perp_market, now_ts);
+    let liqee_positive_settle_limit = liqee_perp_position.available_settle_limit(&perp_market).1;
+
+    // The max settleable amount does not need to be constrained by the liqor's perp settle health,
+    // because taking over perp pnl decreases liqor health: every unit of pnl taken costs
+    // (1-positive_pnl_liq_fee) USDC and only gains init_pnl_asset_weight in perp health.
+    let max_settle = I80F48::from(max_quote_transfer);
+
+    // Take over the liqee's base in exchange for quote
+    let liqee_base_lots = liqee_perp_position.base_position_lots();
+    // Each lot the base position gets closer to 0, the unweighted perp health contribution
+    // increases by this amount.
+    let unweighted_health_per_lot;
+    // -1 (liqee base lots decrease) or +1 (liqee base lots increase)
+    let direction: i64;
+    // Either 1+fee or 1-fee, depending on direction.
+    let fee_factor;
+    if liqee_base_lots > 0 {
+        require_msg!(
+            max_base_transfer >= 0,
+            "max_base_transfer can't be negative when liqee's base_position is positive"
+        );
+
+        // the unweighted perp health contribution gets reduced by `base * price * perp_init_asset_weight`
+        // and increased by `base * price * (1 - liq_fee) * quote_init_asset_weight`
+        let quote_init_asset_weight = I80F48::ONE;
+        direction = -1;
+        fee_factor = cm!(I80F48::ONE - perp_market.base_liquidation_fee);
```

Comment:
so would base_liquidation_fee>= positive_pnl_liquidation_fee in practice?

Reply:
can't say